pkg/virt-handler/vm.go

```diff
@@ -2356,6 +2357,7 @@ func (d *VirtualMachineController) vmUpdateHelperMigrationSource(origVMI *v1.Vir
 
 		err = client.MigrateVirtualMachine(vmi, options)
 		if err != nil {
+			d.finalizeMigration(vmi)
```

Comment:
If the code reaches this block, I do not understand why there is a need to call finalize. It seems to me like this may occur only as a setup failure, to the migration. But if that occurs, finalize should not be called.
Finalize should be called only if the migration already started.

If you remove this line, are there e2e tests which fail? (in other words, do you manage to cover this case?)

Reply:
> If the code reaches this block, I do not understand why there is a need to call finalize. 
It seems to me like this may occur only as a setup failure, to the migration. But if that occurs, finalize should not be called.
> Finalize should be called only if the migration already started.

If the migration fails due to setup failure or while running, we will get to this line.

The way I see it `finalizeMigration` should be called whether a migration if failed or not,
its the logic to run when the migration is finished no matter if it failed or not.
Currently there is no clear point in the code that handle the part when the migration is finished - post migration.
As you may already know, there are multiple points on virt-handler code that indicates the migration is finished,
where on each point the migration state is different state.
With the current implementation I am not sure if its possible to have such place in the code
w/o changing `virt-controller` and `virt-handler` migration logic, which is in my opinion too much work for now.
On the other hand if we had such point, we could call `finalizeMigration` once and not needing this PR at all.
> 
> If you remove this line, are there e2e tests which fail? (in other words, do you manage to cover this case?)

Yes, both tests of the failing path fails:
[Failure while migration is running test](https://github.com/kubevirt/kubevirt/pull/5922/commits/f73036acf559f4cf05f9e1de15e44d1a0b6be667#diffda14ca97e77db6bb275427ac69187a539c58fc297e64519628e5385f9f411c5fR1049)
[Failure during setup test ](https://github.com/kubevirt/kubevirt/pull/5922/commits/a5e55d3aebbaa81a4d80f028a4466b012c4652f1#diff-da14ca97e77db6bb275427ac69187a539c58fc297e64519628e5385f9f411c5fR1024)

Reply:
I'm following the code here, let me know when I missed something:
- `client.MigrateVirtualMachine` call reaches the [virt-launcher](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/cmd-server/server.go#L96).
- It calls [MigrateVmi](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/manager.go#L370) which calls in turn to [startMigration](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/live-migration-source.go#L248).
- All these do some preparations, but do not really detach or prepare the domain for migration. The interesting stuff happen in [asyncMigrate](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/live-migration-source.go#L883) but this one will never return an error, so failure in detachment for example will not be propagated back. Nor the artificial assertion with the [annotation](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/live-migration-source.go#L887).

Reply:
You right asyncMigrate will not return anything (the fact that on the caller there is error handling is bad)
But the errors for the detachment and the artificial assertion with the annotation will propagate to virt-handler.

If an error accrues during asyncMigrate, for example:
detachment fails 
- The caller of detachment is [migrationHelper](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/live-migration-source.go#L806),
returns an error that is passed to the error-channel the migration monitor listens to [[1]](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/live-migration-source.go#L893-L910) [[2]](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/live-migration-source.go#L623) [[3]](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/live-migration-source.go#L657-L660) 

- migration monitor will updated the domain migration metadata [[4]](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/live-migration-source.go#L668).

- vm-controller on virt-handler, updates the VMI `Status.MigrationState` according to the domain migration meta [setMigrationProgressStatus](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-handler/vm.go#L573)
- Now the migraiton failure is reflected on the VMI status that will reconciled by the virt-handler

Reply:
You explained the flow correctly per what I know, but I think the connection to calling `client.MigrateVirtualMachine(vmi, options)` is lost here.
The call to that function will not return an error due to a detachment, migration or other setup failures, it will just cause what you just mentioned and get reflected at some point in the future in the VMI status.
If `client.MigrateVirtualMachine(vmi, options)` returns an error, it means we never reached the setup or the migration at all, we failed much earlier. So why would we call finalize then?

And with that the obvious question raises: If a failure occurs in the setup (at the detach, at the annotation condition, etc) or in the middle of the migration, and the error is populated to the VMI status eventually, why is the filizeMigration called and when.

Reply:
I checked it more thoroughly
Its like you said, when `client.MigrateVirtualMachine` is called we wont catch errors from 
`asyncMigrate`(gorutine) and forwarder including: detachment, migration or other setup failures.

We do catch errors form until `asyncMigrate` is called, for example:
when a migration is already running will get "migration job already executed" error.
The error will propagate and `finalizeMigration` will run.
Here is the path:
`client.MigrateVirtualMachine` ->  `MigrateVirtualMachine` -> `migrateVMI` [[1]](https://github.com/kubevirt/kubevirt/blob/a5e55d3aebbaa81a4d80f028a4466b012c4652f1/pkg/virt-launcher/virtwrap/cmd-server/server.go#L110)  -> 
`startMigration` [[2]](https://github.com/kubevirt/kubevirt/blob/a5e55d3aebbaa81a4d80f028a4466b012c4652f1/pkg/virt-launcher/virtwrap/manager.go#L370) [[3]](https://github.com/kubevirt/kubevirt/blob/a5e55d3aebbaa81a4d80f028a4466b012c4652f1/pkg/virt-launcher/virtwrap/live-migration-source.go#L248) -> `initializeMigrationMetadata` [[4]](https://github.com/kubevirt/kubevirt/blob/a5e55d3aebbaa81a4d80f028a4466b012c4652f1/pkg/virt-launcher/virtwrap/live-migration-source.go#L254) 
and finally we will hit [[5]](https://github.com/kubevirt/kubevirt/blob/a5e55d3aebbaa81a4d80f028a4466b012c4652f1/pkg/virt-launcher/virtwrap/live-migration-source.go#L302) 

Also this code run more then once (as it invoked from vm controller loop),
which means we will always get these errors.
This explains why [failure while migration is running test](https://github.com/kubevirt/kubevirt/blob/f73036acf559f4cf05f9e1de15e44d1a0b6be667/tests/network/vmi_multus.go#L1049) pass's.